sysfs/net_class.go

```diff
@@ -194,6 +256,247 @@ func parseNetClassIface(devicePath string) (*NetClassIface, error) {
+		case "resend_igmp":
+			attrs.ResendIgmp = vp.PInt64()
+		case "tlb_dynamic_lb":
+			attrs.TLBDynamicLB = vp.PInt64()
+		case "updelay":
+			attrs.UpDelay = vp.PInt64()
+		case "use_carrier":
+			attrs.UseCarrier = util.ParseBool(value)
+		case "xmit_hash_policy":
+			attrs.TransmitHashPolicy = &value
+		}
+
+	}
+	return &attrs, nil
+}
+
+func parseNetClassBondDeviceAttrs(devicePath string) (*NetClassBondDeviceAttrs, error) {
+	attrs := NetClassBondDeviceAttrs{}
+	files, err := ioutil.ReadDir(devicePath)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, f := range files {
+		if !f.Mode().IsRegular() {
+			continue
+		}
+		name := filepath.Join(devicePath, f.Name())
+		value, err := util.SysReadFile(name)
+		if err != nil {
+			if os.IsNotExist(err) || os.IsPermission(err) || err.Error() == "operation not supported" || err.Error() == "invalid argument" {
+				continue
+			}
+			return nil, fmt.Errorf("failed to read file %q: %w", name, err)
+		}
+		vp := util.NewValueParser(value)
+		switch f.Name() {
+		case "ad_actor_oper_port_state":
+			attrs.AdActorOperationalPortState = vp.PUInt64()
+		case "ad_aggregator_id":
+			attrs.AdAggregatorId = vp.PUInt64()
+		case "ad_partner_oper_port_state":
+			attrs.AdPartnerOperationalPortState = vp.PUInt64()
+		case "link_failure_count":
+			attrs.LinkFailureCount = vp.PUInt64()
+		case "mii_status":
+			attrs.MiiStatus = util.ParseBool(value)
+		case "perm_hwaddr":
+			mac, err := net.ParseMAC(value)
+			if err != nil {
+				return nil, err
+			}
+			attrs.PermamentHWAddress = &mac
+		case "queue_id":
+			attrs.QueueID = vp.PUInt64()
+		case "state":
+			attrs.State = vp.PUInt64()
+		}
+	}
+	return &attrs, nil
+}
+
+func (fs FS) resolveBondingRelationships(netClass *NetClass) error {
+	for _, netClassIface := range *netClass {
+		if netClassIface.BondAttrs != nil {
+			path := filepath.Join(fs.sys.Path(netclassPath), netClassIface.Name, "bonding")
+			if _, err := os.Stat(filepath.Join(path, "active_slave")); !os.IsNotExist(err) {
+				active_slave, err := util.SysReadFile(filepath.Join(path, "active_slave"))
+				if err != nil {
+					return fmt.Errorf("unable to read %s", filepath.Join(path, "active_slave"))
+				}
+				if len(active_slave) > 0 {
+					if intf, exists := (*netClass)[active_slave]; exists {
+						netClassIface.BondAttrs.ActiveDevice = &intf
+					} else {
+						return fmt.Errorf("unable to find device %s", active_slave)
+					}
+				}
+			}
+			if _, err := os.Stat(filepath.Join(path, "slaves")); !os.IsNotExist(err) {
+				devices, err := util.SysReadFile(filepath.Join(path, "slaves"))
+				if err != nil {
+					return fmt.Errorf("unable to read %s", filepath.Join(path, "slaves"))
+				}
+				for _, device := range strings.Split(devices, " ") {
+					if intf, exists := (*netClass)[device]; exists {
+						netClassIface.BondAttrs.Devices = append(netClassIface.BondAttrs.Devices, &intf)
+					} else {
+						return fmt.Errorf("unable to find device %s", device)
+					}
+				}
```

Comment:
So it turns out the errors that I was seeing trying to test node_exporter against my branch are because they have an incomplete fixture set:
https://github.com/prometheus/node_exporter/blob/b52bf958f8e2d4ed1624b8122d08af4d12da9322/collector/fixtures/sys.ttar#L1094-L1098

Member devices eth1, eth4, and eth5 are not present in the fixtures.  Does it make more sense to fix the upstream fixtures to be complete (either by removing the references to the non-existing devices, or adding dummy fixtures for them), or by not resolving them to pointers here and not returning an error at all?

Reply:
If we're sure that on real systems the member would be present, which I expect, I'd say to update the fixtures.
I'd include the other interfaces if it's not inflating the fixture sizes to much, otherwise remove the references.